### PR TITLE
Fix `selectRecent` memory leak keeping old players in memory

### PR DESCRIPTION
**Pull request by snosenzo**
**User-Facing Changes**
 - fix frequent memory leak where `selectRecent` would keep the players in memory after selecting a new source

**Description**

Because the callback is exposed to the larger PlayerManager closure context it captures the player in it. This function is used in memoized state in `Start.tsx`, but the `OpenDialog` that contains it isn't cleaned up by react on unmount for an unknown reason. `selectSource` does not seem to have the same problem with its usage. As a workaround, moving the definition of the `selectRecent` callback to an outside function will prevent it from keeping the player in its closure context this way.

The `stateNode` holding onto the detached element is the mui-modal in `OpenDialog` which is unmounted, but it's still stored in react internal state functions (ie: `stateNode` and `lastEffect`), which makes me think this might be a react issue. When we update react we should probably check to see whether this is still the case because apparently cleanup in react 18 is more aggressive. 

Retainer stack:
image

To reproduce prior to fix:
(with memory panel open in chrome devtools)
1. use studio default layout
2. Refresh the page and select a local file source
3. Wait for buffering to finish
4. Go to data source tab and select the same local source
5. Take a snapshot in the memory panel
6. See two iterable players in memory, the old IterablePlayer will be the one that has a greater distance, click to inspect its retainers
7. It's primary retainers should resemble the screenshot above

### CODE REVIEW

**Comment by snosenzo** [comment]
part of fixing: https://github.com/foxglove/studio/issues/4824

**Comment by defunctzombie** [comment]
If we move everything in this component to a zustand store do we think that would solve this situation more broadly rather than picking off one-by-one?

File: packages/studio-base/src/components/PlayerManager.tsx [hunk 1]
```diff
@@ -334,3 +313,36 @@ export default function PlayerManager(props: PropsWithChildren<PlayerManagerProp
     </>
   );
 }
+
+// fixes memory leak across > 2 data sources without reloading
```
Comment (defunctzombie) [inline packages/studio-base/src/components/PlayerManager.tsx:317]: This comment does not tell how the memory leak is fixed or what I should avoid doing to re-introduce the memory leak. I'd recommend either improving the comment or moving the structure of state management towards an approach that prevents this kind of _leak_ generally.

Comment (foxymiles) [reply packages/studio-base/src/components/PlayerManager.tsx:317]: Is there another word we can use besides `context` to describe closed over variables? Just skimming this I'm going to think it means a react context.

Comment (jtbandes) [reply packages/studio-base/src/components/PlayerManager.tsx:317]: closure context....clontext?

**Comment by snosenzo** [comment]
> If we move everything in this component to a zustand store do we think that would solve this situation more broadly rather than picking off one-by-one?

That's a good question. Potentially? It's hard to say without really testing. It's probably be worth attempting in the future, but I worry that we still might run into the same problem that the `createPlayerListener` function has in the message pipeline.

File: packages/studio-base/src/components/PlayerManager.tsx (hunk 1 above)
Comment (snosenzo) [reply packages/studio-base/src/components/PlayerManager.tsx]: I added more info to the PR description and updated the comment to be a little more clear. let me know if it's still confusing

**Pull request merged**